### PR TITLE
Fix a group colocation issue

### PR DESCRIPTION
**Pull request by kgaillot**
If colocations are configured with an individual group member rather than the group itself, it can cause incorrect node assignment under certain conditions. This PR ensures that member colocations are always considered.

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , can you review this when you get a chance? Thanks

File: lib/pacemaker/pcmk_sched_group.c [hunk 1]
```diff
@@ -701,18 +723,35 @@ pcmk__group_with_colocations(const pe_resource_t *rsc,
...
                                                      list);
         }
+
+        if (!pe__group_flag_is_set(rsc, pe__group_colocated)) {
+            // @COMPAT Non-colocated groups are deprecated
+            return;
+        }
+
+        // Add explicit colocations involving the group's (other) children
+        for (GList *iter = rsc->children; iter != NULL; iter = iter->next) {
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_group.c:749]: const, optionally (and same for "with this")

File: lib/pacemaker/pcmk_sched_group.c [hunk 2]
```diff
@@ -722,8 +761,7 @@ pcmk__group_with_colocations(const pe_resource_t *rsc,
      * happen, so the group's mandatory colocations must be explicitly added.
      */
     for (GList *iter = rsc->children; iter != NULL; iter = iter->next) {
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_group.c:763]: const, optionally

**Review by nrwahl2 — COMMENTED** [review]
Awesome.

I thought maybe we could add a test for `pcmk__coloc_explicit` to `cmp_colocation_priority()` instead, but that didn't change the results.

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 3]
```diff
@@ -338,21 +338,24 @@ pcmk__new_colocation(const char *id, const char *node_attr, int score,
 {
     pcmk__colocation_t *new_con = NULL;
 
-    if (score == 0) {
-        crm_trace("Ignoring colocation '%s' because score is 0", id);
-        return;
-    }
+    CRM_CHECK(id != NULL, return);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c:332]: Good point, no

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 4]
```diff
@@ -338,21 +338,24 @@ pcmk__new_colocation(const char *id, const char *node_attr, int score,
...
-    if (new_con == NULL) {
+    if (score == 0) {
+        pe_rsc_trace(dependent,
+                     "Ignoring colocation '%s' (%s with %s) because score is 0",
+                     id, dependent->id, primary->id);
         return;
     }
 
+    new_con = calloc(1, sizeof(pcmk__colocation_t));
+    CRM_ASSERT(new_con != NULL);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c:348]: Yeah, there's no ideal answer for that. Adding a return value to `pcmk__new_colocation()` would be a lot of clutter and not recoverable anyway

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 5]
```diff
@@ -372,8 +375,9 @@ pcmk__new_colocation(const char *id, const char *node_attr, int score,
     new_con->node_attribute = pcmk__s(node_attr, CRM_ATTR_UNAME);
     new_con->influence = influence;
 
-    pe_rsc_trace(dependent, "%s ==> %s (%s %d)",
-                 dependent->id, primary->id, new_con->node_attribute, score);
+    pe_rsc_trace(dependent, "Added colocation %s (%s with %s @%s using %s)",
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c:369]: good point

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 6]
```diff
@@ -602,6 +605,10 @@ colocate_rsc_sets(const char *id, xmlNode *set1, xmlNode *set2, int score,
             if (rsc_2 == NULL) {
                 pcmk__config_err("%s: No resource found for %s",
                                  id, xml_rsc_id);
+                // Should be possible only with validation disabled
+                pcmk__config_err("Ignoring resource %s and later in set %s "
+                                 "for colocation with set %s: No such resource",
+                                 xml_rsc_id, set2, set1);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c:613]: Bah, I always misremember the order

**Comment by kgaillot** [comment]
Rebased on current main and added 4 commits on top to address review

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 7]
```diff
@@ -674,9 +674,9 @@ colocate_rsc_sets(const char *id, const xmlNode *set1, const xmlNode *set2,
                                                    xml_rsc_id);
             if (rsc_1 == NULL) {
                 // Should be possible only with validation disabled
-                pcmk__config_err("Ignoring resource %s and later in set %s "
-                                 "for colocation with set %s: No such resource",
-                                 xml_rsc_id, ID(set1), ID(set2));
+                pcmk__config_err("Ignoring colocation of set %s resource %s "
+                                 "with set %s: No such resource",
+                                 ID(set1), xml_rsc_id, ID(set2));
                 return;
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_colocation.c:680]: continue

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 8]
```diff
@@ -938,23 +938,29 @@ pcmk__unpack_colocation(xmlNode *xml_obj, pe_working_set_t *data_set)
...
-    const char *influence_s = crm_element_value(xml_obj,
-                                                XML_COLOC_ATTR_INFLUENCE);
+    const char *score = NULL;
+    const char *influence_s = NULL;
 
-    if (score) {
-        score_i = char2score(score);
+    if (pcmk__str_empty(id)) {
+        pcmk__config_err("Ignoring " XML_CONS_TAG_RSC_DEPEND
+                         " without " CRM_ATTR_ID);
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_colocation.c:950]: Return?

File: lib/pacemaker/pcmk_sched_colocation.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_sched_colocation.c:354]: I guess checking in `pcmk__unpack_colocation()` now (latest update) means that if we fail here, it's from an internal constraint. Looks good.

**Review by nrwahl2 — APPROVED** [review]
Noted two minor things individually, otherwise looks good

File: lib/pacemaker/pcmk_sched_colocation.c (hunk 7 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c:680]: thanks

File: lib/pacemaker/pcmk_sched_colocation.c (hunk 8 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_colocation.c:950]: whoops

**Comment by kgaillot** [comment]
updated commits for latest review

**Review by nrwahl2 — APPROVED** [review]

**Pull request merged**